idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,271 @@
+    gradient_f, gradient_c, col,row, line_dic = get_dfds_dcds(model, theta_names, tee)
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True)))
+    
+    
+    
+    # calculate error propagation of the objective fuction
+    # = df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)
+    # = (df/ds*ds/dp)*cov*(df/ds*ds/dp)
+    # step 1. df/ds*ds/dp 
+    
+    # [1 x (Nx + Np)] matrix
+    df_ds = np.reshape(gradient_f,(1,len(gradient_f)))
+    
+    # [(Nx + Np) x (Np) ] matrix
+    ds_dp = np.reshape(dsdp,(len(dsdp),len(theta_names)))
+    
+    # [1 x Np ] matrix
+    fssp = np.matmul(df_ds,ds_dp)
+    
+    # step 2. (df/ds*ds/dp)*cov*(df/ds*ds/dp).transpose()
+    # [1 x Np] x [Np x Np ] x [Np x 1] = [1 x 1]
+    propagation_f = fssp @ cov_ @ fssp.transpose()
+    
+    assert propagation_f.shape == (1,1), "propagation_f should be a 1 x 1 matrix. Something is wrong if you are seeing this..."
+    
+    # convert to scalar
+    propagation_f = propagation_f[0,0]
+    
+    # calculate error propagation of constraints
+    # = dc/dp*cov_p*dc/dp + (dc/dx*dx/dp)*cov_p*(dc/dx*dx/dp)
+    # = (dc/ds*ds/dp)*cov*(dc/ds*ds/dp)
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True))) 
+    if num_constraints > 0:
+        # gradient_c rearrange.
+        # k_aug sparse form is [col_idx, row_idx, val] with index starts from 1
+        # python sparse form is [row_idx, col_idx, val] with index starts from 0
+        # note: vairable 'row' from get_dfds_dcds includes objecive function name. i.e, Ncon = len(row)-1
+        '''
+        row_idx = gradient_c[:,1]-1
+        col_idx = gradient_c[:,0]-1
+        data = gradient_c[:,2]
+        gradient_c = sparse.csr_matrix((data, (row_idx, col_idx)), shape=(len(row)-1, len(col)))
+        '''
+        # step 1. dc/ds*ds/dp
+        # [Ncon x (Nx + Np)] x [(Nx + Np] x Np] = [Ncon x Np] matrix
+        cssp = np.matmul(gradient_c.toarray(), dsdp)
+        
+        # step 2. (dc/ds*ds/dp)*cov*(dc/ds*ds/dp).transpose()
+        # [Ncon x Np] x [Np x Np] x [Np x Ncon]
+        
+        # Not sure if this is correct.
+        # propagation_c = np.sum(np.matmul(cssp,cov)*cssp,axis=1)
+        
+        # Updated to match propgation_f
+        propagation_c = cssp @ cov_ @ cssp.transpose()
+    else:
+        propagation_c = np.array([])
+    
+    Output = namedtuple('Output',['gradient_f', 'gradient_c', 'dsdp', 'propagation_c','propagation_f','col','row'])
+    results= Output(gradient_f, gradient_c, sparse.csr_matrix(dsdp), propagation_c, propagation_f,col,row)
+    return results
+
+
+def clean_variable_name(theta_names):
```

Comment:
I don't think we should need this function. What is the reason that we must remove single quotes from variable names?

Reply:
We had a similar issue with a string name if the name contains `white spaces` or `'` e.g) [`NRTL example`](https://github.com/IDAES/idaes-pse/blob/5e370d6f803ff35f68c9c0cd5329360bb9c9bc23/idaes/apps/uncertainty_propagation/examples/uncertainty_propagation_NRTL.py#L20). 

For example, when `variable_name = ["fs.properties.tau['benzene', 'toluene']", "fs.properties.tau['toluene','benzene']"]`, `parmest_class.theta_est()` works but `parmest_class.theta_est(calc_cov=True)` fails. 

Please let me know if you have any ideas to resolve this.

Reply:
@Robbybp If my memory is correct, we added this function to deal with a limitation in parmest. On my summer goals is parmest cleanup, which will hopefully eliminate the need for this.

Reply:
Presumably these names are being used to locate a component? I just tested, in the examples directory,
```python
>>> import uncertainty_propagation_NRTL as nrtl
>>> model = nrtl.NRTL_model({"temperature": 365.5})
>>> for name in nrtl.variable_name:
...     model.find_component(name).name
... 
'fs.properties.tau[benzene,toluene]'
'fs.properties.tau[toluene,benzene]'
```
I don't see a problem locating the components with these names, and am worried if the names are being used for anything other than (a) a key in a dictionary or (b) to find their components. Although if it's necessary for Parmest, it's necessary for Parmest, and not something we're going to fix here.